### ISSUE TITLE
Can I detect device's max supported resolution for H.264 Main Profile

### ISSUE THREAD

**Issue by rajeshghosh1978** [google/ExoPlayer#6233]
### [REQUIRED] Searched documentation and issues
I have already searched for this, but not able to locate a specific question.

### [REQUIRED] Question
Right now we are supporting all types of Android devices available in the market. We have added our logic to differentiate between L1 and L3 devices. If it is an L3 device we are restricting to 480X360 **(Based on this link https://developer.android.com/guide/topics/media/media-formats)**.
Since video quality is the key feature for a video app , we are trying to support 1280 x 720 px which is not supported by all the device. So our intention is to add more intelligence to our code, which will determine if the device is capable to render those resolutions before even they start playback.  Is this possible ?

### A full bug report captured from the device
This is a question. For bug report please check my other open issue i.e. https://github.com/google/ExoPlayer/issues/5935

### Link to test content
I can't share any link. All the contents is under vendor rights.

**Comment by rajeshghosh1978** [comment]
Any update on this ?

**Comment by vladimir-kazakov** [comment]
I'm not an Android developer, but as far as I know, it's possible to query [media capabilities](https://developer.android.com/reference/android/media/MediaCodecInfo) of a device. The maximum supported resolution of a codec is defined by its level. You can find this info in H.264 specifications - either [ISO/IEC 14496-10](https://www.iso.org/standard/66069.html) ([available for free](https://standards.iso.org/ittf/PubliclyAvailableStandards/)) or [ITU-T H.264](https://www.itu.int/rec/T-REC-H.264). This info is also available in [Wikipedia](https://en.wikipedia.org/wiki/H.264/MPEG-4_AVC#Levels), but it may not be as precise and as complete as in the specifications.

To play encoded (compressed) media, devices or apps use decoders (decompressors). These decoders are built to support specific codec profiles and levels. Different platforms, including Android, allow to query this info, which allows to identify which media will be playable by an available decoder.

So, in order to solve your problem, you need to identify levels of H.264 that you need to support. 720p resolution, at 30 fps max, is supported starting from the level 3.1. Now, using some Android API, you can query media capabilities of the available decoder. If it supports level 3.1 or higher, you, sort of, can be sure that the playback of 720p media will work. If it supports lower levels, then it's an indication that probably the playback of this resolution won't work.

I'm saying "sort of" and "probably", because it's not a guarantee. The specification tries to make it a guarantee, but not all decoders are implemented in a good way. Sometimes, they even lie about their capabilities. But usually, everything is more or less accurate.

_P.S. This question has nothing to do with ExoPlayer, unless ExoPlayer exposes some API to query media capabilities. I don't know whether it has something like this. I'd guess that it doesn't, because the underlying platform usually can provide this info._

**Comment by marcbaechinger** [comment]
You can get a `MediaCodecInfo` object by using the [`MediaCodecUtil`](https://github.com/google/ExoPlayer/blob/f6297f4f51f00c20d5b937d8f636ef215d89730f/library/core/src/main/java/com/google/android/exoplayer2/mediacodec/MediaCodecUtil.java#L134). It provides you with the capabilities of the codec on a device.

**Comment by rajeshghosh1978** [comment]
I have already tried that out, still, want to make sure whether I am on the right track, I need your expertise. The code I have written so far 




```
public void testGetCapabilities() {
		Log.d(TAG, "testGetCapabilities");
		int codecCount = MediaCodecList.getCodecCount();
		for (int i = 0; i < codecCount; ++i) {
			MediaCodecInfo info = MediaCodecList.getCodecInfoAt(i);
			if (!info.getName().startsWith("OMX.")) {
				Log.d(TAG, "  skipping...");
				continue;
			}
			String[] types = info.getSupportedTypes();
			for (int j = 0; j < types.length; ++j) {
				if(types[j].equalsIgnoreCase(MIME_TYPE_VIDEO)) {
					Log.d(TAG, (i + 1) + ": " + info.getName());
					Log.d(TAG, "  isEncoder = " + info.isEncoder());
					Log.d(TAG, "calling getCapabilitiesForType " + types[j]);
					MediaCodecInfo.CodecCapabilities cap = info.getCapabilitiesForType(types[j]);

					MediaCodecInfo.VideoCapabilities videoCapabilities = cap.getVideoCapabilities();
					Log.d(TAG, "videoCapabilities =  " +videoCapabilities.areSizeAndRateSupported(TEST_WIDTH,TEST_HEIGHT,TEST_FRAME_RATE));
					Range<Integer> height = videoCapabilities.getSupportedHeights();
					Range<Integer> width = videoCapabilities.getSupportedWidthsFor(height.getUpper());

					Log.d(TAG, "max supported width =  " +width.getUpper()+" :: height = "+height.getUpper() );
					Log.d(TAG, "========================================================" );
				}
			}
		}
	}
```

**Comment by ojw28** [comment]
The classes @marcbaechinger referred to above are easier to use than the ones you're looking at. Note that they're part of the ExoPlayer library, not part of the Android SDK directly:

https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/mediacodec/MediaCodecUtil.html
https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/mediacodec/MediaCodecInfo.html

**Comment by rajeshghosh1978** [comment]
I have tried that one, but not able to find out a way to create Format object, since I only have Codec Name and Video width and height as an input.  
@marcbaechinger Do you have any example for creating a Format object.

**Comment by ojw28** [comment]
Why do you need to create a `Format` object? You can use these APIs like this:

```
// Get the MediaCodecInfo for the mime type you're interested in
// (you should also set the second parameter to true if you require a secure decoder)
MediaCodecInfo codecInfo =
    MediaCodecUtil.getDecoderInfo(MimeTypes.VIDEO_H264, false, false);

// Query whatever you want in the MediaCodecInfo. Note you can retrieve the video
// capabilities using the line below, which gets you to where you are near the bottom
// of your code in a more concise way...
VideoCapabilities capabilities = codecInfo.capabilities.getVideoCapabilities();
```

**Comment by rajeshghosh1978** [comment]
Yes, you are right. I was missing the last part. Here is the full code to detect max supported width and height for specific codec

```
public static final String MIME_TYPE_VIDEO = "video/avc";
	public static final String CODEC = "avc1.4d001f";
	
	private void testGetCapabilities(){
		try {
			List<MediaCodecInfo> mediaCodecInfoList = MediaCodecUtil.getDecoderInfos(MIME_TYPE_VIDEO,true,false);

			for(MediaCodecInfo mediaCodecInfo:mediaCodecInfoList) {
				if(mediaCodecInfo.isCodecSupported(CODEC)){
					android.media.MediaCodecInfo.VideoCapabilities videoCapabilities = mediaCodecInfo.capabilities.getVideoCapabilities();

					Range<Integer> height = videoCapabilities.getSupportedHeights();
					Range<Integer> width = videoCapabilities.getSupportedWidthsFor(height.getUpper());

					Log.d(TAG, "max supported width =  " +width.getUpper()+" :: height = "+height.getUpper() );
					Log.d(TAG, "========================================================" );
				}
			}

		} catch (MediaCodecUtil.DecoderQueryException e) {
			e.printStackTrace();
		}
	}
```

**Comment by rajeshghosh1978** [comment]
So when we say codec capabilities, does this means it is the device which is capable to decode the video for the said resolution. Or this is just a capability range for the codec itself. Maybe codec is capable to decode the stream for certain devices.

**Comment by ojw28** [comment]
It's what the codec is capable of doing on the device the code is running on (i.e. it's what you want). By the way, if you deliver your media using DASH or HLS, ExoPlayer will just select the streams that the device can play automatically, and you wouldn't need to worry about doing this yourself.

**Comment by rajeshghosh1978** [comment]
@ojw28 : That's what I assumed before till I raised this ticket #5935 :(

**Comment by ojw28** [comment]
I don't really see how the two issues are related. ExoPlayer is going to be doing pretty much the same thing internally for its stream selection as you're looking at doing externally in this issue. So unless you think you're going to do it better than ExoPlayer in some way (if so then please contribute the change directly into ExoPlayer), you're not going to get a different result and "fix" #5935 by not playing streams that ExoPlayer tries to play. There's also no particular evidence that #5935 is related to attempting to play a stream at an unsupported quality level.

I'll leave this issue closed because I think you know which APIs to look at now. We'll look at #5935 when we have a chance, but it's very unlikely we'll be able to do anything useful without a working test stream.

**Comment by rajeshghosh1978** [comment]
@ojw28 : My intention was to create a fallback approach to reduce the occurrence, without compromising the vie quality. So, please keep this as close, I will wait for the fix.
